contributors/design-proposals/volume-metrics.md

```diff
@@ -0,0 +1,138 @@
+
+### Metric Types
+
+Since we are interested in count(or rate) and time it takes to perform certain volume operation - we will use [Histogram](https://prometheus.io/docs/practices/histograms/) type for
+emitting these metrics.
+
+We will be using `HistogramVec` type so as we can attach dimensions at runtime. Name of operation will become
+part of metric name and at minimum name of volume plugin will be emitted as a dimension.
+
+We are also interested in count of volume operation failures and hence a metric of type `NewCounterVec`
+will be used for keeping track of errors.
+
+Following is a list of metrics (not exhaustive) that will be added by this proposal:
+
+
+```
+storage_volume_attach_seconds { volume_plugin = "aws-ebs" }
+storage_volume_detach_seconds { volume_plugin = "aws-ebs" }
+storage_provision_seconds { volume_plugin = "glusterfs" }
+storage_volume_delete_seconds { volume_plugin = "gce-pd" }
+storage_volume_mount_seconds { volume_plugin = "vsphere" }
+storage_volume_unmount_seconds { volume_plugin = "iscsi" }
+storage_volume_mount_device_seconds { volume_plugin = "aws-ebs"}
+storage_volume_unmount_device_seconds { volume_plugin = "aws-ebs"}
+storage_verify_volume_attach_seconds { volume_plugin = "cinder" }
+```
+
+Similarly errors will be named:
+
+```
+storage_volume_attach_errors_total { volume_plugin = "aws-ebs" }
+storage_volume_detach_errors_total { volume_plugin = "aws-ebs" }
+storage_provision_errors_total { volume_plugin = "glusterfs" }
+storage_volume_delete_errors_total { volume_plugin = "gce-pd" }
+storage_volume_mount_errors_total { volume_plugin = "vsphere" }
+storage_volume_unmount_errors_total { volume_plugin = "iscsi" }
+storage_volume_mount_device_errors_total { volume_plugin = "aws-ebs"}
+storage_volume_unmount_device_errors_total { volume_plugin = "aws-ebs"}
+storage_verify_volume_attach_errors_total { volume_plugin = "cinder" }
+```
+
+### Implementation Detail
+
+We propose following changes as part of implementation details.
+
+1. All volume operations are executed via `goroutinemap.Run` or `nestedpendingoperations.Run`.
+`Run` function interface of these two types can be changed to include a `operationComplete` callback argument.
+
+   For example:
+
+   ```go
+   // nestedpendingoperations.go
+   Run(v1.UniqueVolumeName, types.UniquePodName, func() error, opComplete func(error)) error
+   // goroutinemap
+   Run(string, func() error, opComplete func(error)) error
+   ```
+
+   This will enable us to know when a volume operation is complete.
+
+2. All `GenXXX` functions in `operation_generator.go` should return plugin name in addition to function and error.
+
+   for example:
+
+   ```go
+   GenerateMountVolumeFunc(waitForAttachTimeout time.Duration,
+       volumeToMount VolumeToMount,
```

Comment:
`VolumeToMount` contains volume spec and at that point plugin name is kind of unknown.  The resolution of which plugin will perform mounting or attaching or some other volume operation - usually happens inside `GenXXX` functions of `operationGenerator` module - https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_generator.go#L360

So returning plugin that was chosen to perform certain operation avoid having to modify all these internal structs.